### ISSUE TITLE
[Bug] Variadic placeholder parsing error

### ISSUE THREAD

**Issue by mritd** [caddyserver/caddy#5716]
### The issue description

The `doImport` function in [parse.go](https://github.com/caddyserver/caddy/blob/b32f265eca/caddyconfig/caddyfile/parse.go#L497) seems unable to correctly parse the token, which will ultimately lead to an error in parsing the "Variadic placeholder".

### The running environment

- Caddy: `v2.7.3 h1:eMCNjOyMgB5A1KgOzT2dXKR4I0Va+YHCJYC8HHu+DP0=`
- OS: `Ubuntu Server 22.04`

### The configuration to reproduce the issue

```sh
(PKG_PROXY) {
    reverse_proxy {args[0]}://{args[1]} {
        header_up Host {args[1]}
    }
}

:80 {
    import PKG_PROXY https www.google.com
}
```

### The error log

> Variadic placeholder {args[0]}://{args[1]} has an invalid start index

```sh
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5864484,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":""}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"warn","ts":1691469998.5865088,"logger":"caddyfile","msg":"Variadic placeholder {args[0]}://{args[1]} has an invalid start index","file":"/etc/caddy/Caddyfile:2","import_chain":["/etc/caddy/Caddyfile:8 (import PKG_PROXY)"]}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"warn","ts":1691469998.5871835,"msg":"Caddyfile input is not formatted; run 'caddy fmt --overwrite' to fix inconsistencies","adapter":"caddyfile","file":"/etc/caddy/Caddyfile","line":2}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5877721,"logger":"admin","msg":"admin endpoint started","address":"localhost:2019","enforce_origin":false,"origins":["//localhost:2019","//[::1]:2019","//127.0.0.1:2019"]}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"warn","ts":1691469998.587835,"logger":"http.auto_https","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":80}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5879872,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0x4000248b80"}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5880983,"logger":"http.log","msg":"server running","name":"srv0","protocols":["h1","h2","h3"]}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5881765,"msg":"autosaved config (load with --resume flag)","file":"/var/lib/caddy/.config/caddy/autosave.json"}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5881784,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/var/lib/caddy/.local/share/caddy"}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5882018,"msg":"serving initial configuration"}
Aug 08 12:46:38 arm64 caddy[1309]: {"level":"info","ts":1691469998.5882034,"logger":"tls","msg":"finished cleaning storage units"}
```

**Comment by francislavoie** [comment]
Oh geez, that's an interesting edgecase. It happens specifically because you have a token which starts with `{args[` and ends with `]}` but we didn't take into account that it could be _two_ placeholders with some other text in between. So it's trying to use what's in between (i.e. `0]}://{args[1` :sweat_smile:) as the index. Even more complicated by the fact that your text in between has `:` which is meant as the delimiter for a variadic range (e.g. `1:3` for args 1 through 3).

For now I can suggest two workarounds: pass the full upstream address as an arg to the snippet (i.e. `https://www.google.com`) instead of assembling it from args, or skip passing the scheme and instead configure the transport:

```
transport http {
	tls
}
```

**Comment by mritd** [comment]
Thank you for your reply. My initial expectation was to be able to more conveniently set the `Host` request header; 

if I pass the full URL I need to somehow "extract" the `Host`, and some addresses may not support `https`; I may not have expressed myself very clearly, please see the configuration:

```sh
(PKG_PROXY) {
    @{args[0]} host {args[0]}.pkgs.example.com
    reverse_proxy @{args[0]} {args[1]}://{args[2]} {
        header_up Host {args[2]}
    }
}

*.pkgs.example.com {
    import PKG_PROXY ubuntu http archive.ubuntu.com
}
```

After the issue occurred, I found a more elegant solution by looking through the documentation😄:


```sh
(PKG_PROXY) {
    @{args[0]} host {args[0]}.pkgs.example.com
    reverse_proxy @{args[0]} {args[1]} {
        header_up Host {upstream_hostport}
    }
}

*.pkgs.example.com {
    import PKG_PROXY ubuntu http://archive.ubuntu.com
}
```

**Comment by merrywhether** [comment]
I feel like I'm hitting the same bug. I'm trying to vary just the port on a reverse_proxy and `:` characters seems to getting me too. Simplified, I tried this:

```
(app-proxy) {
	reverse_proxy {args[0]}
}

example.com {
	import app-proxy localhost:3200
}
```
but that errored: `http.log.error	making dial info: upstream {args[0]}: invalid dial address : missing port in address`

I tried this other variation:

```
(app-proxy) {
	reverse_proxy {args[0]}:{args{1}}
}

example.com {
	import app-proxy localhost 3200
}
```
and got another error: `http.log.error	making dial info: upstream {args[0]}:{args{1}}: invalid dial address :}: invalid start port`

This would probably be my ideal config:

```
(app-proxy) {
	reverse_proxy localhost:{args[0]}
}

example.com {
	import app-proxy 3200
}
```
but that yielded a third error: `http.log.error	dial tcp [::1]:0: connect: can't assign requested address`

This works, but it is the one variation that doesn't do what I'm aiming for 😢 
```
(app-proxy) {
	reverse_proxy {args[0]}:3200
}

example.com {
	import app-proxy localhost
}
```

Is this the same underlying issue, or am I hitting something else entirely?
